### PR TITLE
MNT: Add modifier key press handling to macosx backend

### LINKED ISSUE #9837: MacOS: Key modifiers deprecated
### Bug report

**Issue**

A heads up that modifiers like:

https://github.com/matplotlib/matplotlib/blob/813d842994a76b94508d8ec4938f9985ddb86cab/src/_macosx.m#L2601-L2607

are deprecated:

https://developer.apple.com/documentation/appkit/nseventmodifierflags?language=objc

and probably need to be replaced by the flags like `NSEventModifierFlagCommand` etc.  

I'd change, but I'm not sure how this will affect older OS X installs etc...

### LINKED ISSUE #20486: Modifier key press events not recognized on MacOSX backend
### Bug report

**Bug summary**

Key press events identify every key, except modifiers (control, option, command and shift). This behaviour is different from that of Linux and Windows where modifiers are recognized. This behaviour is identical to what was presented in #9835, and supposedly fixed in #9836.   

**Code for reproduction**

```python
import matplotlib.pyplot as plt
fig = plt.figure()

def on_key(event):
    print('you pressed', event.key, event.xdata, event.ydata)

fig.canvas.mpl_connect('key_press_event', on_key)
plt.show()
```

**Actual outcome**

All characters recognized, except modifier keys. 

**Expected outcome**

We should be seeing identifiers, such as `shift` and `control` as output from the above script. 

**Matplotlib version**

  * Operating system: MacOS Big Sur 11.4
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`):  3.4.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 3.9.5
  * Jupyter version (if applicable): n/a
  * Other libraries: n/a

### PR DESCRIPTION
**Pull request by greglucas**
## PR Summary

The flagsChanged event handles single presses of modifier keys, so we need to send the corresponding string to our key press handlers from within that routine. Modifier + second key is handled within the KeyUp/KeyDown routines already, so leave those alone.

Taking the example from issue #20486
```python
import matplotlib.pyplot as plt
fig = plt.figure()

def on_key(event):
    print('you pressed', event.key, event.xdata, event.ydata)
def on_key_release(event):
    print('you released', event.key, event.xdata, event.ydata)

fig.canvas.mpl_connect('key_press_event', on_key)
fig.canvas.mpl_connect('key_release_event', on_key_release)
plt.show()
```
this adds the single modifier key presses back in. Before this, the (modifier + key) "cmd+a" would show up, but there wasn't a single "cmd" (modifier) preceding it.

```bash
you pressed cmd None None
you pressed cmd+a None None
```

I'm not sure how to send key press events through the terminal natively, so there are no automated tests.

closes #20486
closes #9837

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Review by jklymak — APPROVED** [review]
This works as advertised.  Not an expert on macOS backend, so not sure this is the "right" fix, but....

File: src/_macosx.m [hunk 1]
```diff
@@ -2066,29 +2066,31 @@ - (BOOL)acceptsFirstResponder
     return YES;
 }
 
-/* This is all wrong. Address of pointer is being passed instead of pointer, keynames don't
-   match up with what the front-end and does the front-end even handle modifier keys by themselves?
-
+// flagsChanged gets called on single modifier keypresses
+// The modifier + second key gets handled in convertKeyEvent in KeyUp/KeyDown
```
Comment (anntzer) [inline src/_macosx.m:2070]: does it also work for e.g. "ctrl+shift"? (dunno, just asking)

File: src/_macosx.m [hunk 2]
```diff
@@ -2066,29 +2066,31 @@ - (BOOL)acceptsFirstResponder
...
 - (void)flagsChanged:(NSEvent*)event
 {
-    const char *s = NULL;
-    if (([event modifierFlags] & NSControlKeyMask) == NSControlKeyMask)
+    const char* s = NULL;
+    if (([event modifierFlags] & NSEventModifierFlagControl) == NSEventModifierFlagControl)
```
Comment (anntzer) [inline src/_macosx.m:2074]: Also, you can just check the boolean state of the & (without the equality check); this is shorter and consistent with the implementation in convertKeyEvent.

Comment (greglucas) [reply src/_macosx.m]: :+1:, blindly copying examples ;) Not sure why they do it that way in many online examples then. I agree that is simpler and makes sense.

File: src/_macosx.m (hunk 1 above)
Comment (greglucas) [reply src/_macosx.m:2070]: a ctrl, shift, f for example produces this:
```python
you pressed cmd None None
you pressed shift None None
you pressed cmd+F None None
you pressed cmd None None
```

I think the extra command at the end is there due to the shift also being let up and changing state while command is still pressed...

Comment (anntzer) [reply src/_macosx.m:2070]: On other backends, the second press is "cmd+shift", not just shift.  (Well, ctrl+shift on non-osx.)

Comment (greglucas) [reply src/_macosx.m]: OK, so this was more complicated than I had hoped... The flagsChanged gets called on any press OR release, which is why I was printing out "pressed" when it was really a "release".

I think this means we have to keep track of the state of the keys ourselves and what is going up/down. I think this now matches what other backends do, including press/release properly.

File: src/_macosx.m [hunk 3]
```diff
@@ -2066,29 +2097,87 @@ - (BOOL)acceptsFirstResponder
...
+        lastCommand = true;
+        keyChangeCommand = true;
+        keypress = true;
+    }
+    else if (!([event modifierFlags] & NSEventModifierFlagCommand) && lastCommand) {
+        // Command released
+        lastCommand = false;
+        keyChangeCommand = true;
+        keyrelease = true;
+    }
```
Comment (anntzer) [inline src/_macosx.m:2117]: Not a huge deal, but I guess each of these pairs can be grouped to
```objc
bool isPress = true;  // false for release
if ([event modifierFlags] & NSEventModifierFlagCommand) {
    lastCommand = !lastCommand
    keyChangeCommand = true
    isPress = lastCommand;
} else if (...)
    ...
} else {  // Unknown flag
    return;
}
```

Comment (greglucas) [reply src/_macosx.m:2117]: `NSEventModifierFlagCommand` will only be true when it is held down, so we would miss the release with that implementation I think... We could probably squash the two cases into a single if block with your suggestion, but maybe I'm missing a more elegant way to keep track of the push/release here:

```objc
    if ((([event modifierFlags] & NSEventModifierFlagCommand) && !lastCommand) ||
        (!([event modifierFlags] & NSEventModifierFlagCommand) && lastCommand)) {
        // Command pressed or released
        lastCommand = !lastCommand;
        keyChangeCommand = true;
        keypress = lastCommand;
    } ...
```

Comment (anntzer) [reply src/_macosx.m]: Ah, sorry, I missed the inversion.  Perhaps a good place for a `^` (xor)?

Comment (greglucas) [reply src/_macosx.m]: :+1: Yep, a good place for it, thanks.

Comment (anntzer) [reply src/_macosx.m]: LGTM, although probably another mac user should test :-)

**Comment by anntzer** [comment]
Ah, perhaps also update the table at https://github.com/matplotlib/matplotlib/blob/main/doc/users/explain/event_handling.rst#event-connections?

**Comment by greglucas** [comment]
Good suggestion about updating the table.

* The shift + 2 on US keyboards is `@`, so I'm not sure whether the `"` for the other backends are expected or not.
* I don't have an altGr key, so I can't check that.
* I don't have a capital A key, I'm not entirely clear what that entry is referring to? Possibly pressing a with caps_lock on...

**Comment by anntzer** [comment]
> The shift + 2 on US keyboards is @, so I'm not sure whether the " for the other backends are expected or not.

Looks like the original author was from Mexico where shift+2 is " (http://kbdlayout.info/KBDLA/), probably the entry should be fixed for qwerty.

> I don't have an altGr key, so I can't check that.

Per https://en.wikipedia.org/wiki/AltGr_key I guess that's "left alt/option"

> I don't have a capital A key, I'm not entirely clear what that entry is referring to? Possibly pressing a with caps_lock on...

I guess no one has, probably that was capslocked indeed?

**Comment by anntzer** [comment]
Thanks for the table update.  I'll merge in a day or two if the macos crew doesn't manifest themselves for a second review :)

**Comment by jklymak** [comment]
I've not been following. Has it diverged significantly from when I reviewed?  I can try it again.

**Comment by anntzer** [comment]
Mostly just check that things work just as advertised in the table under event_handling.rst.

**Review by anntzer — COMMENTED** [review]
modulo @jklymak checking, per the above.

**Review by anntzer — APPROVED** [review]

**Comment by greglucas** [comment]
There was quite a bit of logic rework since your initial review, and this now handles modifier key_release events too. Mostly just mash some keys and make sure you get back what you're expecting when testing with the script included in the example. If you have an AltGr key, it would be good to update the table entry for that too, as I can't check that.

**Comment by jklymak** [comment]
I've never heard of an AltGr key so I guess if don't have one.  I'll try and check tomorrow

**Review by dstansby — APPROVED** [review]
Works well for me, and fixes issues as advertised. I haven't reviewed the code in `_macosx.m`, but the doc update looks good.

**Comment by dstansby** [comment]
Oh, and I don't think macs have AltGr keys, so leaving that empty is fine.

**Comment by greglucas** [comment]
I am going to self-merge this based on the 3 approvals and to avoid future rebases. It is a bugfix, but it has been broken for a while, so I don't think we should worry about backporting this.

**Pull request merged**